sdk/storage/azure-storage-datamovement/src/job_engine.cpp

```diff
@@ -0,0 +1,466 @@
+            progress.NumFilesSkipped = _internal::AtomicLoad(m_numFilesSkipped);
+            progress.TotalBytesTransferred = _internal::AtomicLoad(m_totalBytesTransferred);
+            m_hydrateParameters.ProgressHandler(progress);
+          }
+        }
+      }
+      if (currNumUndone == 0)
+      {
+        uint32_t partId = jobPart->m_id;
+        jobPart.reset();
+        m_engine->PartDone(m_jobId, partId);
+      }
+    }
+
+    JobPlan::~JobPlan()
+    {
+      if (m_rootTask)
+      {
+        auto currStatus = m_rootTask->SharedStatus->Status.load(std::memory_order_relaxed);
+        if (currStatus == JobStatus::InProgress)
+        {
+          m_rootTask->SharedStatus->Status.compare_exchange_strong(
+              currStatus, JobStatus::Paused, std::memory_order_relaxed, std::memory_order_relaxed);
+        }
+      }
+      m_jobParts.clear();
+
+      if (m_numAliveParts)
+      {
+        while (m_numAliveParts->load(std::memory_order_relaxed) != 0)
+        {
+          std::this_thread::yield();
+        }
+      }
+    }
+
+    JobEngine::JobEngine(const std::string& plansDir, _internal::TransferEngine* transferEngine)
+        : m_plansDir(plansDir), m_transferEngine(transferEngine)
+    {
+      _internal::CreateDirectory(m_plansDir);
+
+      m_messageProcessor = std::thread([this]() {
+        std::unique_lock<std::mutex> guard(m_messageMutex, std::defer_lock);
+        while (true)
+        {
+          guard.lock();
+          m_messageCond.wait_for(guard, std::chrono::milliseconds(100));
+
+          std::vector<EngineOperation> messages;
+          messages.swap(m_messages);
+          guard.unlock();
+
+          for (auto& m : messages)
+          {
+            ProcessMessage(m);
+          }
+
+          if (m_messageProcessorStop.load(std::memory_order_relaxed))
+          {
+            break;
+          }
+
+          const size_t RefillQueueThreshold = 5000;
+          if (m_transferEngine->m_numTasks.load(std::memory_order_relaxed) < RefillQueueThreshold
+              && !m_transferEngine->m_stopped.load(std::memory_order_relaxed))
+          {
+            auto tasks = GetMoreTasks();
+            if (!tasks.empty())
+            {
+              m_transferEngine->AddTasks(std::move(tasks));
+            }
+          }
+        }
+      });
+    }
+
+    JobEngine::~JobEngine()
+    {
+      m_messageProcessorStop.store(true, std::memory_order_relaxed);
+      if (m_messageProcessor.joinable())
+      {
+        m_messageProcessor.join();
+      }
+    }
+
+    JobProperties JobEngine::CreateJob(
+        _internal::JobModel model,
+        _internal::HydrationParameters hydrateParameters)
+    {
+      // TODO: exception handling during job creation/resumption
+      auto uuid = Core::Uuid::CreateUuid().ToString();
+      EngineOperation op;
+      op.JobId = uuid;
+      op.Type = decltype(op.Type)::CreateJob;
+      op.Model = model;
+
+      EngineOperation op2;
```

Comment:
Why need both of create and resume here?

Reply:
_Create_ operation only creates job plan files on disk. It doesn't start the job. A job after creation is in a status as if it was paused. So we need to resume it.